File: Design-Patterns-Using-Ruby/SOLID/DependencyInversionPrinciple.rb
# Define the abstract IMailSender class
class IMailSender
    def send_mail(to_address, from_address, subject, body)
        raise NotImplementedError, 'Subclasses must implement the send_mail method'
    end
end

# Implement the SmtpServer class extending IMailSender
class SmtpServer < IMailSender
    def send_mail(to_address, from_address, subject, body)
        puts "Send mail: subject: #{subject} from: #{from_address} to: #{to_address} body: #{body}"
    end
end

# Implement the EmailSender class
class EmailSender
    def initialize(mail_sender)
        @mail_sender = mail_sender
    end

    def send_email(to_address, from_address, subject, body)
        # Delegate email sending to the mail sender implementation
        @mail_sender.send_mail(to_address, from_address, subject, body)
    end
end

# Client code
# Create an instance of the SmtpServer class
smtp_server = SmtpServer.new

# Create an instance of the EmailSender class and pass in the SmtpServer instance
email_sender = EmailSender.new(smtp_server)

# Send an email using the EmailSender instance
email_sender.send_email(
  'recipient@example.com',
  'sender@example.com',
  'mail subject.',
  'This is a test email body.'
)
=begin 
Send mail: subject: mail subject. from: sender@example.com to: recipient@example.com body: This is a test email body.
=end

File: Design-Patterns-Using-JavaScript/SOLID/DependencyInversionPrinciple.js
class IMailSender {
    sendMail(toAddress, fromAddress, subject, body) {
        // Abstract method, to be implemented by subclasses
    }
}

class SmtpServer extends IMailSender {
    sendMail(toAddress, fromAddress, subject, body) {
        console.log(`Send mail: subject: ${subject} from: ${fromAddress} to: ${toAddress} body: ${body}`);
    }
}

class EmailSender {
    constructor(mailSender) {
        this.mailSender = mailSender;
    }

    sendEmail(toAddress, fromAddress, subject, body) {
        // Delegate email sending to the mail sender implementation
        this.mailSender.sendMail(toAddress, fromAddress, subject, body);
    }
}

// Client code.
const smtpServer = new SmtpServer();
const emailSender = new EmailSender(smtpServer);

// Send an email using the EmailSender instance
emailSender.sendEmail(
    "recipient@example.com",
    "sender@example.com",
    "mail subject.",
    "This is a test email body."
);

/*
Send mail: subject: mail subject. from: sender@example.com to: recipient@example.com body: This is a test email body.
*/
